src/bounded_queue.ml

```diff
@@ -0,0 +1,113 @@
+(*
+  structure node t value: data type, next: pointer to node t
+  structure queue t Head: pointer to node t, Tail: pointer to node t, H lock: lock type, T lock: lock type
+*)
+
+type 'a node = Nil | Next of 'a * 'a node ref
+
+type 'a t = {
+  mutable head : 'a node ref;
+  mutable tail : 'a node ref;
+  head_lock : Mutex.t;
+  tail_lock : Mutex.t;
+  not_empty : Condition.t;
+  not_full : Condition.t;
+  size : int Atomic.t;
+  capacity : int;
+}
+
+(* create and initialize the 2 lock queue *)
+let create () =
+  let dummy = ref Nil in
+  {
+    head = dummy;
+    tail = dummy;
+    head_lock = Mutex.create ();
+    tail_lock = Mutex.create ();
+    not_empty = Condition.create ();
+    not_full = Condition.create ();
+    size = Atomic.make 0;
+    capacity = 1_000_000;
+  }
+
+(* push element to tail of queue *)
+let push t value =
+  let new_tail = ref Nil in
+  let new_node = Next (value, new_tail) in
+  Mutex.lock t.tail_lock;
+  (* Ensures queue is not full *)
+  while Atomic.get t.size = t.capacity do
+    Condition.wait t.not_full t.tail_lock
+  done;
+  (* Check if queue is empty before this insertion *)
+  t.tail := new_node;
+  t.tail <- new_tail;
+  let empty = Atomic.fetch_and_add t.size 1 = 0 in
```

Comment:
okay sure, the equality check at line 68 too could be moved outside critical section since similar logic right, or is there any reason why you didn't comment on that.

Reply:
> equality check at line 68 too could be moved outside critical section

Yes, exactly.

> any reason why you didn't comment on that

I don't want to repeat and I think it is good a practise, as a developer, to look for more instances of any observation.